### ISSUE TITLE
Roadmap

### ISSUE THREAD

**Issue by voldemortX** [voldemortX/pytorch-auto-drive#4]
Roadmap for our users (to state feature requests) and contributors.

*\* Low priority tasks.*

**2021Q2:**
- Lane detection methods
    - [ ] Add RESA (VGG16, ResNets)
    - [ ] Explore ERFNet-RESA
    - [ ] Add ResNet18-LSTR
    - [ ] Add ResNet34-LSTR/ERFNet-LSTR
    - [ ] Add ERFNet-PRNet
    - [ ] *Add ENet-SAD
    - [ ] *Explore ERFNet-SAD
- Semantic segmentation methods
    - [ ] *Add ERFNet training from scratch
- Datasets and pre-processings
    - [ ] *Support general affine transforms for keypoints
    - [ ] *Support BDD100K
    - [ ] Support LLAMAS
- Visualization (#7)
    - [ ] *Support demo with video input for lane detection
    - [ ] *Support demo with video input for semantic segmentation
    - [ ] *Support demo with camera input for lane detection
    - [ ] *Support demo with camera input for semantic segmentation
- Benchmark
    - [ ] Investigate the "ENet slower than ERFNet" problem
    - [ ] Count fps/flops/mem for transformer-based method LSTR
- Documentation
    - [ ] *Explanations/Descriptions for re-implemented methods, especially the improved parts
- Framework
    - [ ] *Support multi-GPU training

**2021Q1:**
- Lane detection methods
    - [x] Add ResNet backbones  cea2ce8
    ~~Add RESA (VGG16, ResNets)~~
    ~~Explore ERFNet-RESA~~
    ~~Add ResNet18-LSTR~~
    ~~Try add a LSTR that is directly comparable with other methods, i.e. on a common backbone~~
    ~~Add ERFNet-PRNet~~
    - [x] Add ENet Baseline ed3f739
    ~~Add ENet-SAD~~
    ~~*Explore ERFNet-SAD~~
- Semantic segmentation methods
    - [x] Add and test ENet on Cityscapes ee3444b
    ~~*Add ERFNet training from scratch~~
    - [x] Add --workers option 89d3695
- Datasets and pre-processings
    - [x] Support TuSimple and CULane loading as keypoints & keypoint transforms (Rotation, Resize)  #5, 839d096
    ~~Support BDD100K~~
    ~~*Support LLAMAS~~
- Visualization
    - [x] Add lane markers visualization toolkit for images 6b29f02 (only support visualization from files )
    - [x] Organize segmentation result visualization toolkit for images 5ed43d9
- Benchmark
    - [x] *Explore FPS tests 93b2a21
    - [x] *Try provide FLOPs and memory counts for implemented methods 93b2a21
- Documentation
    - [x] Better guide for downloading and preparing datasets (partly addressed by #8) 
    - [x] *Guide for visualization toolkits c680c90

**Comment by voldemortX** [comment]
The maintainers have rather low bandwidth these days, about half features for Q1 remain unfinished and are pushed to Q2. Any help would be much appreciated!

**Comment by SikandAlex** [comment]
I am interested in helping to support this library. I am in a position to add support for comma10k as well as GAN-based weather augmentations soon.

**Comment by voldemortX** [comment]
@SikandAlex It will be an honor to have your help as well!

> I am interested in helping to support this library. I am in a position to add support for comma10k as well as GAN-based weather augmentations soon. I have previously trained this on BDD100K: https://github.com/hustvl/YOLOP

Any new supports on datasets is welcomed!

> I would also like to move these two papers into the Roadmap:
> 
> https://arxiv.org/abs/2105.05003 CondLaneNet https://arxiv.org/abs/2105.13680 FOLOLane
> 

The CondLaneNet is open-sourced and could be easier to implement. While FOLOLane might prove a harder method that need more work, since we do not yet have one of its backbones (BiSeNet).

I'll add them in the Roadmap for Q4 and they can of course continue into 22. You can submit PRs whenever you have a ready-to-go bunch of codes (e.g. implemented one of the dataset class and tested its loading, or finished an algorithm).
Thanks again for your help!

**Comment by voldemortX** [comment]
TensorRT support is also what @cedricgsh and I have talked about recently. We too agree that pytorch-auto-drive should not stop at a research codebase. Our primal aim would be a TensorRT benchmark for model speed and op-based FLOPs calculation from fvcore. But given our current bandwidth, I think that would need to wait until 22Q1 at least (a refactor of the framework might be required).

**Comment by SikandAlex** [comment]
I have an AGX Xavier on hand that I will hopefully be able to provide some benchmarks on for certain models. Unfortunately I'm no expert at TensorRT custom layers etc and some operations it seems are unable to be supported by many developers.

There is so many papers that claim to get certain FPS on deploy to embedded GPUs but they never release their code, they only release testing code and no training code, the results are not reproducible even if there is training code, so really it is a huge mess. I have spent the past 2 months trying to determine the best papers and best approach as of Fall 2021 given my computational limitations.

After reading as many papers as I can over the past 3 months, driving around in my friend's Tesla, seeing many other models (amazing AI research from Asia is just destroying us here in US in my opinion), there are not that many critical components to Level 2 highway autonomy (which should be the first step). 

1) Object Detection for Vehicles/Pedestrians 

- YOLOX by Megvii (won AI City Challenge for Streaming Perception for Autonomous Driving 2021, lot of hype, good for edge)
- YOLOV5 by Ultralytics (great support and community, I have extensive experience)
- Yolo-FastestV2 https://github.com/dog-qiuqiu/Yolo-FastestV2 (super lightweight variant)

2) Segmentation of Driveable Surface / Road

After much research I narrowed down the candidate models to the following:

- SFNet
- DDRNet (I successfully train model that does inference at 100+ FPS on RTX 2080 Ti)
- DF-Seg
- Attanet
- STDC

3) Lane Detection (Segmentation with Post-Processing OR Polynomial/Keypoint/Row-Wise/Transformer/Non-Segmentation etc)

I have found that real-time segmentation approaches in regard to lane markings need to incorporate high-resolution features and not resize the input image before doing anything as many networks do. This is because the semantics of the lane lines beyond a certain distance from the vehicle are lost at lower resolution and then you can't predict the path far enough in advance. I also do not know how to post-process the segmentation based approach properly as I have not yet experimented with DBSCAN or RANSAC. From all my readings, this shows the most potential to me but it seems to require the 4-lane probability output rather than just the binary mask for lane segmentation I am currently producing:

https://github.com/czming/RONELD-Lane-Detection
https://arxiv.org/abs/2010.09548

When I tested the models at https://github.com/Turoad/lanedet#Benchmark-and-model-zoo
the CondLaneNet model performed the best which is why I recommended it be placed on the road map although it uses a heavy Resnet-101 backbone.

4) Depth (Monocular/Stereo, ideally monocular) 

- PyDNet
- MobileStereoNet
- MiDaS
- MonoDepth(2/Wavelet)
- FastDepth

5) 3D Object Detection (Maybe can replace Depth)

- FCOS3D
- FCOS3D++/PGD
- DD3D

6) Bird's Eye View / Top View / Projection Transform 

Finally, this is the best Github project that I've been able to find related to self-driving. All the models seem to come from OpenVINO / PINO model library. 

https://github.com/iwatake2222/self-driving-ish_computer_vision_system

On the control side of things, MPC solver is solution to use for latitude/longitude control.

This is the extensive information I have been able to collect. I automatically discarded models that didn't have code implementations but I think I could have made a mistake here or there. I was going to keep all this information to myself but all I do is code non-stop all day and have no life and still only make slow progress. My friends just work at company's like Facebook and Google and don't wan't to work on anything exciting. Impossible to get them to give up stock to work with me, also has a learning curve. So hopefully by giving back to the open-source community, they will give back to me and we can all improve science and also make money.

**Comment by voldemortX** [comment]
Well my knowledge about self-driving is kind of only in the research stage for now. And I certainly learned a lot from your comments. Though I'm still skeptical about deep learning's performance in actual applications of self-driving, especially on RGB inputs. So now what I do in SenseTime is more about human-centric applications.

Btw, we'll release our own lane detector later, perhaps early 22 at the latest. It end-to-end achieves a reasonable performance (~75 CULane, ~95 LLAMAS) at 150 FPS in PyTorch with a small model, which we believe could be beneficial to application.

**Comment by voldemortX** [comment]
@SikandAlex We have now pushed initial supports for ONNX and TensorRT conversions, maybe they can be helpful to your applications? Refer to [DEPLOY.md](https://github.com/voldemortX/pytorch-auto-drive/blob/master/docs/DEPLOY.md).
